### PR TITLE
Variant annotation with variant sets

### PR DESCRIPTION
**Pull request by david4096**
This PR attempts to formalize the relationship between variant sets and variant annotation sets. Previously, it was possible to have variant annotations for variants that weren't present in GA4GH. Now annotation VCFs are considered a source of a variant set as well as a variant annotation set.

This is done by placing a small description JSON next to the VCF in the directory. For now, this file simply denotes whether the VCF is annotated or not.

    {
        "isAnnotated": true
    }

Furthermore, the requirement of placing the variant object within the annotation has been removed: annotations now maintain the start and end coordinates themselves so they can be searched across without looking up the variant.

### CODE REVIEW

File: ga4gh/datamodel/datasets.py [hunk 1]
```diff
@@ -210,11 +210,10 @@ def __init__(self, localId, dataDir, backend):
                     self, localId, relativePath, backend)
                 self.addVariantSet(variantSet)
             # Variant annotations sets
-            variantAnnotationSetDir = os.path.join(
-                relativePath, "variantAnnotations")
-            if os.path.isdir(variantAnnotationSetDir):
+            if self.isAnnotated(relativePath, localId):
                 variantAnnotationSet = variants.HtslibVariantAnnotationSet(
-                    self, localId, variantAnnotationSetDir, backend)
+                        self, localId, relativePath, backend,
```
Comment (jeromekelleher) [inline ga4gh/datamodel/datasets.py]: I think this is over-indented for hanging indent --- newer versions of pep8 might choke on this.

File: ga4gh/datamodel/datasets.py [hunk 2]
```diff
@@ -227,6 +226,27 @@ def __init__(self, localId, dataDir, backend):
                     self, localId, bamPath, backend)
                 self.addReadGroupSet(readGroupSet)
 
+    def descriptionExists(self, fname):
+        # check for existence of description JSON
```
Comment (jeromekelleher) [inline ga4gh/datamodel/datasets.py]: Use docstring comments for public methods. i.e.
```python
def descriptionExists(self, fname):
   """
   Check for existence of description JSON.
   """
```

File: ga4gh/datamodel/datasets.py [hunk 3]
```diff
@@ -227,6 +226,27 @@ def __init__(self, localId, dataDir, backend):
...
+        return os.path.exists(fname)
+
+    def parseDescriptionJson(self, str):
+        # parse into python dict
+        return json.loads(str)
+
+    def descriptionFilename(self, path, localId):
```
Comment (jeromekelleher) [inline ga4gh/datamodel/datasets.py]: We'd usually use ``getDescriptionFilename`` here, to make it clearer that your returning the filename.

File: ga4gh/datamodel/datasets.py [hunk 4]
```diff
@@ -227,6 +226,27 @@ def __init__(self, localId, dataDir, backend):
...
+        return os.path.join(path, localId + ".json")
+
+    def isAnnotated(self, path, localId):
+        ret = False
+        fname = self.descriptionFilename(path, localId)
+        if self.descriptionExists(fname):
```
Comment (jeromekelleher) [inline ga4gh/datamodel/datasets.py]: I'm not sure the ``descriptionExists`` does much for readablility here.How about
```
fname = self.getDescriptionFilename(path, localId)
if os.path.exists(fname):
    # etc.
```

File: ga4gh/datamodel/variants.py [hunk 5]
```diff
@@ -592,16 +592,19 @@ class HtslibVariantAnnotationSet(HtslibVariantSet):
     Class representing a single variant annotation set backed by a directory of
     indexed VCF or BCF files.
     """
-    compoundIdClass = datamodel.VariantAnnotationSetCompoundId
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: This is where it should be. It's a class variable, so should be assigned at the class level.

**Comment by jeromekelleher** [comment]
Good incremental progress here @david4096, +1 modulo the minor niggles above. 

We should try to get variation_annotation branch updated against master before merging this as there is a minor conflict with a recent refactor. Also, please squash the commits when rebasing, this is really one minor change rather than 4 or 5.

File: ga4gh/datamodel/datasets.py (hunk 4 above)
Comment (diekhans) [inline ga4gh/datamodel/datasets.py]: `Description' is confusing terminology.  It's generally though
of as natural language text.  Metadata would be much clearer
name.  Please change.

**Comment by jeromekelleher** [comment]
Looks like the merge with didn't go quite according to plan here @david4096. It might be worthwhile rebasing off upstream/varation_annotation now and re-resolving the conflicts.

**Comment by david4096** [comment]
Rebased @jeromekelleher :)

File: ga4gh/datamodel/variants.py [hunk 6]
```diff
@@ -592,16 +592,19 @@ class HtslibVariantAnnotationSet(HtslibVariantSet):
...
-
-    def __init__(self, parentContainer, localId, dataDir, backend):
+    def __init__(self, parentContainer, localId, dataDir,
+                 backend, variantSetId):
         super(HtslibVariantAnnotationSet, self).__init__(
             parentContainer, localId, dataDir, backend)
+        self.compoundIdClass = datamodel.VariantAnnotationSetCompoundId
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: We still need to move this back to the class level.

File: ga4gh/datamodel/variants.py [hunk 7]
```diff
@@ -719,6 +716,17 @@ def getVariantAnnotationId(self, gaVariant, gaAnnotation):
             gaVariant.start, md5)
         return str(compoundId)
 
+    def getVariantId(self, gaVariant):
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: Isn't this an exact duplication of VariantSet.getVariantId?

File: ga4gh/datarepo.py [hunk 8]
```diff
@@ -23,6 +24,8 @@ def __init__(self):
         self._referenceSetIdMap = {}
         self._referenceSetNameMap = {}
         self._referenceSetIds = []
+        self._ontologyNameMap = {}
```
Comment (jeromekelleher) [inline ga4gh/datarepo.py]: Why ontologyNames? We should follow the same pattern as datasets and referenceSets unless there's a good reason not to.

File: ga4gh/datarepo.py [hunk 9]
```diff
@@ -42,6 +45,14 @@ def addReferenceSet(self, referenceSet):
         self._referenceSetNameMap[referenceSet.getLocalId()] = referenceSet
         self._referenceSetIds.append(id_)
 
+    def addOntologies(self, ontologies):
```
Comment (jeromekelleher) [inline ga4gh/datarepo.py]: Why not addOntology(self, ontology)? Again, we should follow the established convention here closely for consistency.

File: ga4gh/datamodel/variants.py (hunk 6 above)
Comment (david4096) [inline ga4gh/datamodel/variants.py]: Hey @jeromekelleher , did you try this change? It is breaking for me and I'm not sure why. It causes a "not enough fields for compound ID" exception.

File: ga4gh/datamodel/variants.py (hunk 7 above)
Comment (david4096) [inline ga4gh/datamodel/variants.py]: No, as it uses the variant set's compound ID as opposed to the annotation set's.

File: ga4gh/datarepo.py (hunk 8 above)
Comment (david4096) [inline ga4gh/datarepo.py]: I think `name` captures the functionality better, but I'm open to your suggestion for naming this variable.

**Comment by jeromekelleher** [comment]
Sorry, I haven't had a chance to review this properly today @david4096. Where are we with this and the other PR? I'll try and look at it tomorrow and give some better feedback.

**Comment by david4096** [comment]
@jeromekelleher I included the suggestions you had made, except as noted above. I also added the ability to search annotation sets, which had been left out.

File: ga4gh/datarepo.py (hunk 8 above)
Comment (macieksmuga) [inline ga4gh/datarepo.py]: This makes perfect sense to me, it's in direct analogy to `referenceSetNameMap` and `datasetNameMap`: Ontologies are pulled up by their name, so there isn't a case (yet) for an `ontologyIdMap`.

**Comment by jeromekelleher** [comment]
OK, let's merge this much so. I think the ontologies stuff is confusing, but we can deal with that as we go.

+1 after rebase on upstream/variation_annotation.

**Comment by diekhans** [comment]
Ontologies are a work in progress through the schema.  They will
continue to evolve based on actual use.

**Pull request merged**